Assets/Scripts/Api/MessageHandling/MessageSender.cs

```diff
@@ -0,0 +1,103 @@
+    /// </summary>
+    public class MessageSender
+    {
+        /// <summary>
+        /// Thread-safe queue to place messages, which will be picked up by the send thread and sent to the server.
+        /// </summary>
+        public readonly BlockingCollection<ISendableMessage> SendQueue = new();
+
+        private readonly ClientWebSocket _connection;
+        private Thread _sendThread;
+
+        private readonly Dictionary<Type, string> _messageIdMap = new();
+
+        public MessageSender(ClientWebSocket connection)
+        {
+            _connection = connection;
+        }
+
+        /// <summary>
+        /// Spawns a thread that continuously listens for messages on the WebSocket connection.
+        /// </summary>
+        public void StartSendingMessages()
+        {
+            _sendThread = new Thread(SendMessagesFromQueue);
+        }
+
+        /// <summary>
+        /// Registers the given message type, with the corresponding message ID, as a message that the client expects to
+        /// be able to send to the server.
+        /// </summary>
+        public void RegisterSendableMessage<TMessage>(string messageId)
+            where TMessage : ISendableMessage
+        {
+            _messageIdMap.Add(typeof(TMessage), messageId);
+        }
+
+        /// <summary>
+        /// Continuously takes messages from the send queue, serializes them and sends them to the server.
+        /// </summary>
+        /// <remarks>
+        /// Implementation based on https://www.patrykgalach.com/2019/11/11/implementing-websocket-in-unity/.
+        /// </remarks>
+        private async void SendMessagesFromQueue()
+        {
+            while (true)
+            {
+                if (_connection.State != WebSocketState.Open)
+                {
+                    Task.Delay(50).Wait();
+                    continue;
+                }
+
+                while (!SendQueue.IsCompleted)
+                {
+                    var message = SendQueue.Take();
+                    var serializedMessage = SerializeToJson(message);
+
+                    await _connection.SendAsync(
+                        serializedMessage,
+                        WebSocketMessageType.Text,
+                        true,
+                        CancellationToken.None
+                    );
+                }
+            }
+        }
+
+        /// <summary>
+        /// Serializes the given message to JSON, wrapping it with the appropriate message ID according to its type.
+        /// </summary>
+        /// <exception cref="ArgumentException">
+        /// If a message ID could not be found for the type of the message. Likely because the message type has not been
+        /// registered with <see cref="RegisterSendableMessage{TMessage}"/>.
+        /// </exception>
+        private byte[] SerializeToJson(ISendableMessage message)
+        {
+            string messageId;
```

Comment:
I deliberately avoided this inlining, as I think it negatively affects readability.

See the example here:
```
if (!_messageIdMap.TryGetValue(message.GetType(), out var messageId))
{
    throw new ArgumentException($"Unrecognized type of message object: '{message.GetType()}'");
}

// Using `messageId` here
```

It would not be immediately obvious to me where the `messageId` variable comes from. In addition, when using it in conjunction with a guard clause like this, it is not obvious that the scope of `messageId` goes beyond the end of the `if` statement. I'm not the only one confused by this: https://stackoverflow.com/questions/44981889/scope-of-variable-with-out-parameter

One alternative is to do away with the guard clause, like this:
```
if (_messageIdMap.TryGetValue(message.GetType(), out var messageId))
{
    var messageJson = new JObject(new JProperty(messageId, message));
    var messageString = messageJson.ToString();
    var messageBytes = Encoding.UTF8.GetBytes(messageString);
    return messageBytes;
}

throw new ArgumentException($"Unrecognized type of message object: '{message.GetType()}'");
```

But then Rider of course complains that we should invert the if statement. I prefer keeping the guard clause as-is, and making the declaration and scope of `messageId` explicit by keeping it on its own line like now. However, this is all opinion, and I'm open to other arguments.